For HCA return only the cloud native URL

mikebaumann opened the pull request:
Currently, all Human Cell Atlas (HCA) is open access, and
no user-specific access control is performed.
The HCA checkout buckets, from which the data is accessed,
grant object read permission to GROUP_All_Users@firecloud.org,
so no additional credentials are required or available.
Therefore, for HCA data access, return only
the cloud-native URL, and bypass the request to Bond.
Both the `martha_v2` and `fileSummaryV1` endpoints have been updated to support this behavior.

The new `getSignedUrlV1` endpoint has been updated to obtain a user pet SA key from SAM for HCA DRS URIs.

The unit tests have been augmented to cover the "happy path" through all functional code changes, and most other paths are already covered by existing tests.

For more information, please see: [Terra/Cromwell Access to Files Identified by DRS URIs for the HCA/Terra MVP](https://docs.google.com/document/d/18WDYr3uS82Eu1GhHmeH6PX7tFY6Zk8piKd3ixOoE8Bg)

Resolves DataBiosphere/azul#982.
Partially resolves DataBiosphere/azul#1090, as changes to Cromwell are also needed for that one.

Commit:
For HCA return only the cloud native URL

Currently, all Human Cell Atlas (HCA) is open access, and
no user-specific access control is performed.
The HCA checkout buckets, from which the data is accessed,
grant object read permission to GROUP_All_Users@firecloud.org,
so no additional credentials are required or available.
Therefore, for HCA data access, return only
the cloud native URL, and bypass the request to Bond.

No tests are included for this change in this commit.

dsde-jenkins commented:
Can one of the admins verify this patch?

panentheos commented on common/helpers.js:
```diff
@@ -76,6 +76,7 @@ function dataObjectUriToHttps(dataObjectUri) {
 const BondProviders = Object.freeze({
     FENCE: 'fence',
     DCF_FENCE: 'dcf-fence',
+    HCA: 'hca', // Human Cell Atlas
```
This implies that there is a bond provider for HCA, which according to the description there isn't. It might be clearer to remove this, return `null` from `determineBondProvider`, and make the bond call conditional on the existence of the provider string.

gpolumbo-broad replied:
At first glance, I agreed with you @panentheos , but now that I had a chance to talk this over with @abaumann and @geoffjentry , I'm OK with it.  Eventually, HCA will have a real provider.  Due to limitations in the DOS spec at the moment, it's hard to know which provider to use for which DOS URIs.  This at least keeps all that logic isolated to Martha.

mikebaumann replied:
@panentheos, @gplumbo-broad Thank you both for your review and comments.
Indeed, this is just a first, tiny step towards a bona fide Bond provider for HCA that will be developed over time. The significance of the Bond provider to the Terra/HCA interaction is clearly evident in this design document [Terra/Cromwell Access to Files Identified by DRS URIs for the HCA/Terra MVP](https://docs.google.com/document/d/18WDYr3uS82Eu1GhHmeH6PX7tFY6Zk8piKd3ixOoE8Bg)

The immediate objective is to enable Terra to access HCA for the HCA handoff to Terra MVP initiative. Currently, all HCA data is open/public access, and anyone can obtain a presigned URL for any data. However, Cromwell currently requires cloud-native URLs, so HCA is providing special access to the Broad (Mint and Terra) via cloud-native URLs.  HCA will be supporting controlled access data with user-specific authN/authZ over time and is developing the Fusillade system to support this.
Although currently, HCA data is open/public access, the bucket containing it is access controlled. For the immediate/near-term, the recent plan had been to develop a minimal HCA provider based on the `mock-provider` to provide a single/fixed HCA service account key for all Terra users. Then @abaumann proposed the simpler, easier short-term solution of simply adding  `GROUP_All_Users@firecloud.org` with read access to the ACL of the HCA bucket, ... and hence this PR. :)

panentheos commented on martha_v2/martha_v2.js:
```diff
@@ -11,7 +11,9 @@ function parseRequest(req) {
-    if (req && req.headers && req.headers.authorization) {
+    // Currently HCA data access does not require additional credentials.
+    // The HCA checkout buckets allow object read access for GROUP_All_Users@firecloud.org.
+    if (req && req.headers && req.headers.authorization && provider != BondProviders.HCA) {
```
In addition to here, we'll also need a similar change in `fileSummaryV1/service_account_keys.js`

gpolumbo-broad replied:
Good catch @panentheos, thanks.  @mikebaumann - I think we're looking at around this line: https://github.com/mikebaumann/martha/blob/764ec6a1bb3f39349165b113facb0c2932069297/fileSummaryV1/service_account_keys.js#L5 and doing something similar which is if the provider is `HCA`, then just return a resolved Promise.

mikebaumann replied:
Thanks, @panentheos, @gpolumbo-broad!
I'm back in the office now and committed the corresponding change for File Summary V1. Please have a look. If you would like the conditional to be structured differently, please let me know.

gpolumbo-broad reviewed (approved):
@dvoet - FYI this PR.  Had a meeting with Jeff and Alex and we talked this through.  I can fill you in if you have additional questions.

Commit:
Incorporate Codacy recommendation

Changed `!=` to `!==` per Codacy recommentation

gpolumbo-broad commented on martha_v2/martha_v2.js:
```diff
@@ -11,7 +11,9 @@ function parseRequest(req) {
-    if (req && req.headers && req.headers.authorization) {
+    // Currently HCA data access does not require additional credentials.
+    // The HCA checkout buckets allow object read access for GROUP_All_Users@firecloud.org.
+    if (req && req.headers && req.headers.authorization && provider !== BondProviders.HCA) {
```
https://github.com/mikebaumann/martha/blob/764ec6a1bb3f39349165b113facb0c2932069297/fileSummaryV1/service_account_keys.js#L5 

We need to do something similar which is if the provider is HCA, then just return a resolved Promise.

This will ensure that behavior in both endpoints, `martha_v2` and `fileSummaryV1` remain consistent.

mikebaumann replied:
This change has now been made, as per my recent comment above.

Commit:
Changed File Summary v1 to bypass Bond for HCA

File Summary V2 was updated to not call Bond for DRS URIs
from the Human Cell Atlas (HCA).
This is consistent with the changes made to the `martha_v2` endpoint.
Currently, all HCA data is open/public access, and the HCA
bucket provides object read access to
`GROUP_All_Users@firecloud.org`.
When HCA starts supporting controlled access data, the plan is to
create an actual Bond provider for HCA.

The HCA DRS response does not include the following
field values that are typically returned by File Summary V1:
`timeCreated`, `updated`, `md5Hash`, `signedUrl`.

Commit:
Address Codacy issues

Commit:
Added unit tests for expected HCA behavior

Added basic unit tests for the currently expected
HumanCellAtlas behavior for determineBondProvider,
martha_v2 and fileSummaryV1.

These tests cover the "happy path" through all
functional code changes in this PR.

No updates were made to integration, live or
smoketests.

Commit:
Changed test for undefined due to Codacy rule

mikebaumann commented:
@gpolumbo-broad Unit tests have been added for all HCA-related functional changes to Martha. 
I think this PR is now done and ready to merge, pending your review.
Thanks!

panentheos commented:
There's one more change needed here. There was a new endpoint added on another branch (`getSignedUrlV1`) that will also need to know about the HCA special case. (A rebase should pick up the new code)

@gpolumbo-broad Note that this is necessary because `determineBondProvider` doesn't return null.

mikebaumann commented:
Thanks, @panentheos. Yes, I think the change you made will be easy to update for the HCA case, and once that is done, it should work as intended/expected for HCA.

Will you be adding tests for the change you made? If so, I will update those as well.

Rebasing my branch pulled in your changes just fine. I expect to update the HCA PR for your latest change later today.

Commit:
Changed getSignedUrlV1 to not call Bond for HCA

Made the same type of change to the new getSignedUrlV1
endpoint as was made for the martha_v2 and fileSummaryV1
endpoints.
For HCA data object URIs, at this point, bypass
the call to Bond, and instead get a user pet SA key from SAM to
sign the URL.

mikebaumann commented:
I pushed a commit to make the necessary HCA-related change for `getSignedUrlV1`.
I then started looking at adding unit tests for this change, and 
a question arose.
@panentheos, regarding your change for getSignedUrlV1, it seems moving the existing `fileSummaryV1.createSignedGsUrl` code into `common.helpers.js` and using that to
implement `handlers.getSignedUrlV1` would reduce code duplication, help ensure
code consistency, and make mocking for testing getSignedUrlV1 easier.
Is there a reason you didn't do that?
Although I suppose there is a way to mock the following in ava/sinon:
```
storage.bucket(bucket).file(object).getSignedUrl
```
mocking the following single call seems a lot easier: :-)
```
createSignedGsUrl
```

panentheos commented:
@mikebaumann The bulk of `createSignedGsUrl` is just plumbing, and the only abstraction of any interest is the expiration timestamp. And while happen to be using the same timestamp in `getSignedUrlV1`, there's no requirement that they be the same as each other going forward. So to me it doesn't seem substantial enough to justify the cost of tying together two otherwise independent functions.

Commit:
Added some basic unit tests for `getSignedUrlV1`

Added unit tests for `getSignedUrlV1` to verify
that the service account key used to sign the URL
is obtained from the appropriate source, SAM or Bond,
depending on the DRS URI value, if any.

Minor changes were required to `getSignedUrlV1.js`
to enable testing with sinon (AFAIK), and these changes
are consistent with similar code in the
`fileSummaryV1` and `martha_v2` endpoints.

mikebaumann commented:
@panentheos Yes, the bulk of `createSignedGsUrl` is just plumbing, yet it happens to be the
same plumbing as is used in `getSignedUrlV1`. That's fine, it is a borderline case for refactoring,
and my main motivation for asking was to simplify unit testing of `getSignedUrlV1`.
Happily, I figured out how to test the current code using ava/sinon, although minor changes were required to `getSignedUrlV1.js` to enable testing with sinon (AFAIK), and these changes are consistent with similar code in the `fileSummaryV1` and `martha_v2` endpoints.

Thanks in advance for your review.

mikebaumann commented:
@gpolumbo-broad, @panentheos Please review this PR again.
The new `getSignedUrlV1` endpoint has been updated for HCA, and basic unit tests were added for the same.
I believe all comments have been addressed/resolved, and hopefully, this will be the final review.

panentheos commented on handlers/getSignedUrlV1.js:
```diff
@@ -1,17 +1,17 @@
+            await apiAdapter.getJsonFrom(`${samBaseUrl()}/api/google/v1/user/petServiceAccount/key`, auth);
         const storage = new Storage({ credentials });
-        const [url] = await storage.bucket(bucket).file(object).getSignedUrl({
+        const url = await storage.bucket(bucket).file(object).getSignedUrl({
```
This doesn't look right. The library call returns a single element array that needs to be destructured.

mikebaumann replied:
Thanks! True, indeed:
```
export declare type GetSignedUrlResponse = [string];
export interface GetSignedUrlCallback {
    (err: Error | null, url?: string): void;
}
```

Albeit unexpected, at least to me.
I restored the original line in 'getSignedUrlV1` and updated the unit test mock value to be an array of string instead of string.

Commit:
Corrected for GCS getSignedUrl returning array

Restored original line in `getSignedUrlV1.js` and
updated unit test mock value for `getSignedUrlV1`
to be an array.

panentheos reviewed (commented):
Looks good to me, I'll let someone on CA give the official approval.

mikebaumann commented:
Thanks very much for the review, @panentheos!
